### PR TITLE
Populate hyperopt defaults using schema

### PR DESCRIPTION
**Pull request by arnavgarg1**
This PR adds a new step to ensure that all of the defaults for search algorithm, split, executor etc. are correctly populated using the schema object before accessing them to prevent key errors. All of these parameters are required by hyperopt and expected to be present downstream for config validation, executor creation, and trial execution.

The reason we've not noticed this before in our tests is because all of our tests specify `search_alg` in the hyperopt config. 

Co-authored-by: Kabir Brar <kabir@predibase.com>
Co-authored-by: Travis Addair <tgaddair@gmail.com>

### CODE REVIEW

File: ludwig/hyperopt/run.py [hunk 1]
```diff
@@ -223,6 +224,11 @@ def hyperopt(
 
     hyperopt_config = full_config[HYPEROPT]
 
+    # Convert hyperopt config to hyperopt schema to populate with schema defaults
+    # This fills in missing splits, executor config, search_alg, etc.
+    hyperopt_config_from_schema = HyperoptConfig.Schema().load(hyperopt_config)
+    hyperopt_config = hyperopt_config_from_schema.to_dict()
```
Comment (justinxzhao) [inline ludwig/hyperopt/run.py:230]: Would the ModelConfig class, e.g. `model_config.py::ModelConfig::_set_hyperopt_defaults()` be a more appropriate place for this code?

Comment (arnavgarg1) [reply ludwig/hyperopt/run.py:230]: That's reasonable, I can move it inside there.

File: tests/integration_tests/test_hyperopt.py [hunk 2]
```diff
@@ -557,3 +557,34 @@ def test_hyperopt_nested_parameters(csv_filename, tmpdir, ray_cluster_7cpu):
...
+            "parameters": {
+                "trainer.learning_rate": {
+                    "lower": 0.0001,
+                    "upper": 0.01,
+                    "space": "loguniform",
+                }
+            },
+            "goal": "minimize",
+            "output_feature": output_features[0]["name"],
+            "validation_metrics": "loss",
```
Comment (abidwael) [inline tests/integration_tests/test_hyperopt.py:584]: I've also seen `metric` being used in some hyperopt configs (e.g. https://ludwig.ai/latest/configuration/hyperparameter_optimization/). Which one is the official/most recent parameter name?

Comment (ksbrar) [reply tests/integration_tests/test_hyperopt.py:584]: Interesting. @arnavgarg1 note that `validation_metrics` is not currently on the hyperopt schema.

Comment (arnavgarg1) [reply tests/integration_tests/test_hyperopt.py:584]: Good catch. The reason the test still works is because of backward compatibility. In either case, I've changed this to be `metric` because that is the latest. Apologies for the confusion (copied the config from another test and then modified it)

File: tests/integration_tests/test_hyperopt_ray.py [hunk 3]
```diff
@@ -207,7 +208,9 @@ def run_hyperopt_executor(
 def test_hyperopt_executor(scenario, csv_filename, tmpdir, ray_cluster_4cpu):
     search_alg = scenario["search_alg"]
     executor = scenario["executor"]
-    run_hyperopt_executor(search_alg, executor, csv_filename, tmpdir)
+    # When using the hb_bohb scheduler, num_epochs must equal max_t (which is 81 by default)
```
Comment (justinxzhao) [inline tests/integration_tests/test_hyperopt_ray.py:211]: How did you come to discover this? Is this worth adding as an auxiliary validation?

**Review by justinxzhao — APPROVED** [review]

File: ludwig/schema/model_config.py [hunk 4]
```diff
@@ -511,7 +513,11 @@ def _set_hyperopt_defaults(self):
         if not self.hyperopt:
             return
 
-        scheduler = self.hyperopt.get("executor", {}).get("scheduler")
+        # Convert hyperopt config to hyperopt schema to populate with schema defaults
+        # This fills in missing splits, executor config, search_alg, etc.
+        self.hyperopt = HyperoptConfig.Schema().load(self.hyperopt).to_dict()
```
Comment (tgaddair) [inline ludwig/schema/model_config.py:518]: Should be:

```
HyperoptConfig.from_dict(..).to_dict()
```

Otherwise we circumvent the class cache layer and leak memory.

**Review by jeffkinnison — APPROVED** [review]

File: ludwig/schema/model_config.py (hunk 4 above)
Comment (ksbrar) [reply ludwig/schema/model_config.py:518]: @tgaddair You mean `ModelConfig.from_dict`, right?

File: tests/integration_tests/test_hyperopt_ray.py (hunk 3 above)
Comment (arnavgarg1) [reply tests/integration_tests/test_hyperopt_ray.py:211]: I didn't know about this either, but found out that we've added this logic here in our ModelConfig object creation logic: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/schema/model_config.py#L532

File: ludwig/schema/model_config.py [hunk 5]
```diff
@@ -511,7 +511,7 @@ def _set_hyperopt_defaults(self):
         if not self.hyperopt:
             return
 
-        scheduler = self.hyperopt.get("executor", {}).get("scheduler")
+        scheduler = self.hyperopt.get("executor", {}).get("scheduler", {})
```
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply ludwig/schema/model_config.py:520]: I did this initially because I found that it was possible for the `scheduler` key to not be inside the `executor` dict, but that's no longer the case once we populate `self.hyperopt` from the `HyperoptConfig` so I will remove it

[the remark this answers is not in the document]
Comment (tgaddair) [reply ludwig/schema/model_config.py:520]: Even if `scheduler` is not in the dict, it will just return `None`.
